File: client/src/app/core/models/status.ts
import { EntityType } from './entity-type';

export enum StatusCategory {
  new = 0,
  backlog = 1,
  todo = 2,
  active = 3,
  done = 4,
  inactive = 5,
}

export const statusCategoryLabels: Record<StatusCategory, string> = {
  [StatusCategory.new]: 'new',
  [StatusCategory.backlog]: 'Backlog',
  [StatusCategory.todo]: 'Todo',
  [StatusCategory.active]: 'Active',
  [StatusCategory.done]: 'Done',
  [StatusCategory.inactive]: 'Inactive',
};

export const statusCategoryOptions = [
  StatusCategory.new,
  StatusCategory.backlog,
  StatusCategory.todo,
  StatusCategory.active,
  StatusCategory.done,
  StatusCategory.inactive,
];

export interface Status {
  id: number;
  workspaceId: number;
  entityType: EntityType;
  name: string;
  key: string;
  description?: string | null;
  color?: string | null;
  sortOrder: number;
  category: StatusCategory;
  isSystem: boolean;
}

export interface CreateStatusRequest {
  entityType: EntityType;
  name: string;
  description?: string | null;
  color?: string | null;
  category: StatusCategory;
}

export interface UpdateStatusRequest extends CreateStatusRequest {
  id: number;
}

export interface ReorderStatusesRequest {
  entityType: EntityType;
  statusIds: number[];
}

File: server/Netptune.Core/Entities/Status.cs
using System.Text.Json.Serialization;

using Netptune.Core.BaseEntities;
using Netptune.Core.Colors;
using Netptune.Core.Enums;
using Netptune.Core.ViewModels.Statuses;

namespace Netptune.Core.Entities;

public record Status : WorkspaceEntity<int>
{
    private string? _color;

    public EntityType EntityType { get; set; }

    public string Name { get; set; } = null!;

    public string Key { get; set; } = null!;

    public string? Description { get; set; }

    public string? Color
    {
        get => _color;
        set => _color = NamedColors.Normalize(value);
    }

    public double SortOrder { get; set; }

    public StatusCategory Category { get; set; }

    public bool IsSystem { get; set; }

    [JsonIgnore]
    public ICollection<ProjectTask> ProjectTasks { get; } = new HashSet<ProjectTask>();

    public StatusViewModel ToViewModel()
    {
        return new()
        {
            Id = Id,
            WorkspaceId = WorkspaceId,
            EntityType = EntityType,
            Name = Name,
            Key = Key,
            Description = Description,
            Color = Color,
            SortOrder = SortOrder,
            Category = Category,
            IsSystem = IsSystem,
        };
    }

}
